compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -449,6 +485,55 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                .tcx
+                                .inherent_impls(adt.did)
+                                .iter()
+                                .copied()
+                                .filter(|def_id| {
+                                    if let Some(assoc) =
+                                        self.associated_item(*def_id, item_name, Namespace::ValueNS)
+                                    {
+                                        // Check for both mode is the same so we avoid suggesting
+                                        // incorrect associated item.
+                                        match (mode, assoc.fn_has_self_parameter, source) {
+                                            (Mode::MethodCall, true, SelfSource::MethodCall(_)) => {
+                                                // We check that the suggest type is actually
+                                                // different from the received one
+                                                // So we avoid suggestion method with Box<Self>
+                                                // for instance
+                                                self.tcx.at(span).type_of(*def_id) != actual
+                                                    && self.tcx.at(span).type_of(*def_id) != rcvr_ty
+                                            }
+                                            (Mode::Path, false, _) => true,
+                                            _ => false,
+                                        }
+                                    } else {
+                                        false
+                                    }
+                                })
+                                .collect::<Vec<_>>();
+                            if inherent_impls_candidate.len() > 0 {
+                                inherent_impls_candidate.sort();
+                                inherent_impls_candidate.dedup();
+                                let type_candidates = inherent_impls_candidate
+                                    .iter()
+                                    .map(|impl_item| self.tcx.at(span).type_of(*impl_item))
+                                    .collect::<Vec<_>>();
+                                // number of type to shows at most.
+                                let limit = if type_candidates.len() == 4 { 4 } else { 3 };
+                                for ty in type_candidates.iter().take(limit) {
+                                    err.note(&format!("the {item_kind} was found for {}", ty));
+                                }
+                                if type_candidates.len() > limit {
+                                    err.note(&format!(
+                                        "the {item_kind} was found for {} more types",
+                                        type_candidates.len() - limit
+                                    ));
+                                }
```

Comment:
Instead of emitting a note on each iteration, construct a single string to be emitted.